js/meshopt_encoder.js

```diff
@@ -0,0 +1,182 @@
+		}
+		var write = 0;
+		for (var i = 0; i < data.length; ++i) {
+			result[write++] = (result[i] < 60) ? wasmpack[result[i]] : (result[i] - 60) * 64 + result[++i];
+		}
+		return result.buffer.slice(0, write);
+	}
+
+	function assert(cond) {
+		if (!cond) {
+			throw new Error("Assertion failed");
+		}
+	}
+
+	function reorder(indices, vertices, optf) {
+		var sbrk = instance.exports.sbrk;
+		var ip = sbrk(indices.length * 4);
+		var rp = sbrk(vertices * 4);
+		var heap = new Uint8Array(instance.exports.memory.buffer);
+		var indices8 = new Uint8Array(indices.buffer, indices.byteOffset, indices.byteLength);
+		heap.set(indices8, ip);
+		if (optf) {
+			optf(ip, ip, indices.length, vertices);
+		}
+		var unique = instance.exports.meshopt_optimizeVertexFetchRemap(rp, ip, indices.length, vertices);
+		// heap may have grown
+		heap = new Uint8Array(instance.exports.memory.buffer);
+		var remap = new Uint32Array(vertices);
+		new Uint8Array(remap.buffer).set(heap.subarray(rp, rp + vertices * 4));
+		indices8.set(heap.subarray(ip, ip + indices.length * 4));
+		sbrk(ip - sbrk(0));
+
+		for (var i = 0; i < indices.length; ++i)
+			indices[i] = remap[indices[i]];
+
+		return [remap, unique];
+	}
+
+	function encode(fun, bound, source, count, size) {
+		var sbrk = instance.exports.sbrk;
+		var tp = sbrk(bound);
+		var sp = sbrk(count * size);
+		var heap = new Uint8Array(instance.exports.memory.buffer);
+		heap.set(new Uint8Array(source.buffer, source.byteOffset, source.byteLength), sp);
+		var res = fun(tp, bound, sp, count, size);
+		var target = new Uint8Array(res);
+		target.set(heap.subarray(tp, tp + res));
+		sbrk(tp - sbrk(0));
+		return target;
+	}
+
+	function maxindex(source) {
+		var result = 0;
+		for (var i = 0; i < source.length; ++i) {
+			var index = source[i];
+			result = result < index ? index : result;
+		}
+		return result;
+	}
+
+	function index32(source, size) {
+		assert(size == 2 || size == 4);
+		if (size == 4) {
+			return new Uint32Array(source.buffer, source.byteOffset, source.byteLength / 4);
+		} else {
+			var view = new Uint16Array(source.buffer, source.byteOffset, source.byteLength / 2);
+			return new Uint32Array(view); // copies each element
+		}
+	}
+
+	function filter(fun, source, count, stride, bits, insize) {
+		var sbrk = instance.exports.sbrk;
+		var tp = sbrk(count * stride);
+		var sp = sbrk(count * insize);
+		var heap = new Uint8Array(instance.exports.memory.buffer);
+		heap.set(new Uint8Array(source.buffer, source.byteOffset, source.byteLength), sp);
+		fun(tp, count, stride, bits, sp);
+		var target = new Uint8Array(count * stride);
+		target.set(heap.subarray(tp, tp + count * stride));
+		sbrk(tp - sbrk(0));
+		return target;
+	}
+
+	return {
+		ready: promise,
+		supported: true,
+		reorderMesh: function(indices, triangles, optsize) {
+			var optf = triangles ? (optsize ? instance.exports.meshopt_optimizeVertexCacheStrip : instance.exports.meshopt_optimizeVertexCache) : undefined;
```

Comment:
Is this intended? Setting `triangles=true, optsize=true` works as expected, but `triangles=true, optsize=false` results in a broken triangle mesh.

Reply:
One thing I recommend with this entire integration in general is to check each component independently:

- You can encode the mesh without using filters and without reordering
- You can reorder the mesh without using encoding
- After that you can separately apply filters if need be :)

Are you seeing broken meshes with just reorder being called? If you can share your code I'll be happy to debug it - there are unit tests for this but they may be incomplete, or the calling code may be subtly wrong.

Reply:
Ah sorry, found the bug (my mistake) halfway through setting up the fiddle. Yes, I'm starting with reordering alone, since it could be useful as a separate step in glTF-Transform.